src/netaddress.cpp

```diff
@@ -328,28 +480,65 @@ enum Network CNetAddr::GetNetwork() const
     return m_net;
 }
 
+static std::string IPv6ToString(Span<const uint8_t> a) {
+    assert(a.size() == ADDR_IPV6_SIZE);
+    return strprintf("%x:%x:%x:%x:%x:%x:%x:%x",
+                     a[0] << 8 | a[1], a[2] << 8 | a[3],
+                     a[4] << 8 | a[5], a[6] << 8 | a[7],
+                     a[8] << 8 | a[9], a[10] << 8 | a[11],
+                     a[12] << 8 | a[13], a[14] << 8 | a[15]);
+}
+
 std::string CNetAddr::ToStringIP() const
 {
-    if (IsTor())
-        return EncodeBase32(m_addr) + ".onion";
-    if (IsInternal())
+    switch (m_net) {
+    case NET_IPV4:
+    case NET_IPV6: {
+        CService serv(*this, 0);
+        struct sockaddr_storage sockaddr;
+        socklen_t socklen = sizeof(sockaddr);
+        if (serv.GetSockAddr((struct sockaddr*)&sockaddr, &socklen)) {
+            char name[1025] = "";
+            if (!getnameinfo((const struct sockaddr*)&sockaddr, socklen, name,
+                             sizeof(name), nullptr, 0, NI_NUMERICHOST))
+                return std::string(name);
+        }
+        if (m_net == NET_IPV4) {
+            return strprintf("%u.%u.%u.%u", m_addr[0], m_addr[1], m_addr[2], m_addr[3]);
+        }
+        return IPv6ToString(m_addr);
+    }
+    case NET_ONION:
+        switch (m_addr.size()) {
+        case ADDR_TORV2_SIZE:
+            return EncodeBase32(m_addr) + ".onion";
+        case ADDR_TORV3_SIZE: {
+
+            uint8_t checksum[torv3::CHECKSUM_LEN];
+            torv3::Checksum(m_addr, checksum);
+
+            // onion_address = base32(PUBKEY | CHECKSUM | VERSION) + ".onion"
```

Comment:
1046e17c nit suggestion if you retouch: s/onion_address/tor v3 onion address/

Reply:
I copied that verbatim from https://gitweb.torproject.org/torspec.git/tree/rend-spec-v3.txt#n2135, but anyway - I now prefixed it with "TORv3".